ogre2/src/Ogre2SegmentationCamera.cc

```diff
@@ -0,0 +1,647 @@
+
+      /// \brief User Data Key to set the label
+      private: std::string labelKey = "label";
+
+      /// \brief Background & unlabeled objects label id in semantic map
+      private: int backgroundLabel = 0;
+
+      /// \brief Background & unlabeled objects color in the colored map
+      private: math::Color backgroundColor;
+
+      /// \brief Segmentation Type (Semantic / Instance)
+      private: SegmentationType type {SegmentationType::Semantic};
+
+      /// \brief True to generate colored map
+      /// False to generate labels ids map
+      private: bool isColoredMap;
+
+      /// \brief Keep track of num of instances of the same label
+      /// Key: label id, value: num of instances
+      private: std::map<int, int> instancesCount;
+
+      /// \brief keep track of the random colors
+      /// key: encoded key of r,g,b values. value: always True
+      private: std::map<int64_t, bool> takenColors;
+
+      /// \brief keep track of the labels which is already colored
+      /// usful for coloring items in semantic mode in LabelToColor()
+      /// key: label id. value: always True
+      private: std::map<int64_t, bool> coloredLabel;
+
+      /// \brief Pseudo num generator to generate colors from label id
+      private: std::default_random_engine generator;
+
+      /// \brief Ogre2 Scene
+      private: Ogre2ScenePtr scene;
+
+      friend class Ogre2SegmentationCamera;
+    };
+    }
+  }
+}
+
+/////////////////////////////////////////////////
+SegmentationMaterialSwitcher::SegmentationMaterialSwitcher(Ogre2ScenePtr _scene)
+{
+  this->scene = _scene;
+  this->backgroundColor.Set(0, 0, 0);
+
+  // plain material to switch item's material
+  Ogre::ResourcePtr res =
+    Ogre::MaterialManager::getSingleton().load("ign-rendering/plain_color",
+        Ogre::ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME);
+
+  this->plainMaterial = res.staticCast<Ogre::Material>();
+  this->plainMaterial->load();
+
+  // plain overlay material
+  this->plainOverlayMaterial =
+      this->plainMaterial->clone("plain_color_overlay");
+  if (!this->plainOverlayMaterial->getTechnique(0) ||
+      !this->plainOverlayMaterial->getTechnique(0)->getPass(0))
+  {
+    ignerr << "Problem creating selection buffer overlay material"
+        << std::endl;
+    return;
+  }
+  Ogre::Pass *overlayPass =
+      this->plainOverlayMaterial->getTechnique(0)->getPass(0);
+  Ogre::HlmsMacroblock macroblock(*overlayPass->getMacroblock());
+  macroblock.mDepthCheck = false;
+  macroblock.mDepthWrite = false;
+  overlayPass->setMacroblock(macroblock);
+}
+
+/////////////////////////////////////////////////
+SegmentationMaterialSwitcher::~SegmentationMaterialSwitcher()
+{
+}
+
+/////////////////////////////////////////////////
+bool SegmentationMaterialSwitcher::IsTakenColor(math::Color _color)
+{
+  // get the int value of the 24 bit color
+  int64_t colorId = _color.R() * 255 * 255 + _color.G() * 255 + _color.B();
```

Comment:
Is this supposed to be a unique ID? Is there any chance that 2 different RGB combinations could yield the same ID with this formula?

Reply:
Yes, we have 3 channels each is 8 bit, that ID gets the decimal value of that 24 bit numbers, and there is no change to have the same ID twice

btw, this is the same way the panoptic is  handled in the official panoptic API proposed by COCO dataset (which introduced the panoptic format)

check that if you want to make sure
https://github.com/cocodataset/panopticapi/blob/master/panopticapi/utils.py#L30